### PR TITLE
HPCC-24265 Azure AD has different default administrators group

### PR DESCRIPTION
**Pull request by mayx**
- Add "AAD DC Administrators" to default permissions
- Don't show the deprecated "Administrators" group in list of available
  default groups or permissions
- Set default SysUser basedn properly for Azure

Signed-off-by: Yanrui Ma <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by mayx** [comment]
@RussWhitehead @afishbeck please review...

File: system/security/LdapSecurity/ldapconnection.hpp [hunk 1]
```diff
@@ -209,6 +212,7 @@ interface ILdapConfig : extends IInterface
     virtual int getMaxConnections() = 0;
     virtual void setResourceBasedn(const char* rbasedn, SecResourceType rtype = RT_DEFAULT) = 0;
     virtual int getLdapTimeout() = 0;
+    virtual bool isAzureAD()  = 0;
```
Comment (RussWhitehead) [inline system/security/LdapSecurity/ldapconnection.hpp:215]: overly pedantic, but extra space before equal sign :)

File: system/security/LdapSecurity/permissions.cpp [hunk 2]
```diff
@@ -1456,9 +1456,25 @@ CSecurityDescriptor* PermissionProcessor::createDefaultSD(ISecUser * const user,
...
+    {
+        MemoryBuffer gmb;
+        lookupSid(AAD_ADMINISTRATORS_GROUP, gmb, GROUP_ACT);
+        psid = (PSID)(gmb.toByteArray());
+        if (psid != nullptr)
+        {
+            rc = AddAccessAllowedAce(pacl, ACL_REVISION, sec2ldap(DEFAULT_ADMINISTRATORS_PERMISSION), psid);
+            if (rc == 0)
+            {
+                int error = GetLastError();
```
Comment (RussWhitehead) [inline system/security/LdapSecurity/permissions.cpp:1469]: intermediate variable not needed

**Review by RussWhitehead — COMMENTED** [review]
A couple very minor comments

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 3]
```diff
@@ -2499,6 +2504,8 @@ class CLdapClient : implements ILdapClient, public CInterface
                 filter.append("uid=").append(act_name);
 
             basedn = m_ldapconfig->getUserBasedn();
+            if (strieq(act_name, m_ldapconfig->getSysUser()))
```
[the remark this answers is not in the document]
Comment (mayx) [reply system/security/LdapSecurity/ldapconnection.cpp:2507]: This is to fix an issue I came across when I tested the case when sysuser and user basedns are different. I'll add "if azure" to make sure zero impact on legacy.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 4]
```diff
@@ -3828,9 +3835,12 @@ class CLdapClient : implements ILdapClient, public CInterface
...
             managedBy.append("");
             descriptions.append("");
-            groups.append("Administrators");
-            managedBy.append("");
-            descriptions.append("");
+            if (!m_ldapconfig->isAzureAD())
+            {
+                groups.append("Administrators");
+                managedBy.append("");
+                descriptions.append("");
```
[the remark this answers is not in the document]
Comment (mayx) [reply system/security/LdapSecurity/ldapconnection.cpp:3842]: They're not redunant (the vars are string arrays, not string buffers...)

**Comment by mayx** [comment]
@RussWhitehead changes made per comments, please review again...

**Comment by RussWhitehead** [comment]
@mayx Looks good

**Review by afishbeck — APPROVED** [review]
@mayx looks fine.

**Pull request merged**